### PR TITLE
Support unaligned array join.

### PR DESCRIPTION
**Pull request by amosbird**
For left ARRAY JOIN, expand to the greatest size. If all sizes are zero,
resize to one which is the same as the old behavior, i.e. emptyArrayToSingle

For non-left ARRAY JOIN, expand to the greatest size but keep empty if all sizes are zero.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/src/Interpreters/ExpressionActions.cpp [hunk 1]
```diff
@@ -144,8 +144,9 @@ ExpressionAction ExpressionAction::arrayJoin(const NameSet & array_joined_column
     a.array_joined_columns = array_joined_columns;
     a.array_join_is_left = array_join_is_left;
 
-    if (array_join_is_left)
-        a.function_builder = FunctionFactory::instance().get("emptyArrayToSingle", context);
+    a.function_builder = FunctionFactory::instance().get("length", context);
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/ExpressionActions.cpp]: Strange name.

File: dbms/src/Interpreters/ExpressionActions.cpp [hunk 2]
```diff
@@ -373,26 +374,39 @@ void ExpressionAction::execute(Block & block, bool dry_run) const
...
-                any_array_ptr = non_empty_array_columns.begin()->second;
-                if (ColumnPtr converted = any_array_ptr->convertToFullColumnIfConst())
-                    any_array_ptr = converted;
+            for (const auto & name : array_joined_columns)
+            {
+                auto & src_col = block.getByName(name);
 
-                any_array = &typeid_cast<const ColumnArray &>(*any_array_ptr);
+                Block tmp_block{src_col, column_of_max_length, {{}, src_col.type, {}}};
+                function_arrayResize->build({src_col, column_of_max_length})->execute(tmp_block, {0, 1}, 2, rows);
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/ExpressionActions.cpp]: How it will affect performance?

File: dbms/src/Interpreters/ExpressionActions.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/src/Interpreters/ExpressionActions.cpp]: Better to use separate variable.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Performance test is required.

**Comment by amosbird** [comment]
The performance degrade badly for small arrays. It makes sense to add an option then.


```
Running: array_join
[
{
    "hostname": "201.nobida.cn",
    "main_metric": "max_rows_per_second",
    "num_cores": 20,
    "num_threads": 40,
    "ram": 134768111616,
    "runs": [
        {
            "max_rows_per_second": 67098782.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2] b FROM system.numbers) AS t ARRAY JOIN a, b WHERE NOT ignore(a + b)"
        },
        {
            "max_rows_per_second": 33748463.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2] b FROM system.numbers) AS t LEFT ARRAY JOIN a, b WHERE NOT ignore(a + b)"
        },
        {
            "max_rows_per_second": 21724193.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2] b FROM system.numbers) AS t ARRAY JOIN a, b WHERE NOT ignore(a + b) SETTINGS enable_unaligned_array_join = 1"
        },
        {
            "max_rows_per_second": 21570510.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2] b FROM system.numbers) AS t LEFT ARRAY JOIN a, b WHERE NOT ignore(a + b) SETTINGS enable_unaligned_array_join = 1"
        },
        {
            "max_rows_per_second": 16701070.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2, number] b FROM system.numbers) AS t ARRAY JOIN a, b WHERE NOT ignore(a + b) SETTINGS enable_unaligned_array_join = 1"
        },
        {
            "max_rows_per_second": 16921843.000000,
            "query": "SELECT count() FROM (SELECT [number] a, [number * 2, number] b FROM system.numbers) AS t LEFT ARRAY JOIN a, b WHERE NOT ignore(a + b) SETTINGS enable_unaligned_array_join = 1"
        }
    ],
    "server_version": "18.14.9",
    "test_name": "array_join",
    "time": "2018-12-12 11:11:32"
}
]
```

**Comment by alexey-milovidov** [comment]
Ok. A setting looks reasonable.

**Pull request merged**